### PR TITLE
[InstCombine] Folding `(icmp eq/ne (and X, -P2), INT_MIN)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for folding `(icmp eq/ne (and X, -P2), INT_MIN)`; NFC**
- **[InstCombine] Folding `(icmp eq/ne (and X, -P2), INT_MIN)`**

Folds to `(icmp slt/sge X, (INT_MIN + P2))`

Proofs: https://alive2.llvm.org/ce/z/vpNFY5

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LG

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -5015,6 +5015,22 @@ Instruction *InstCombinerImpl::foldICmpBinOp(ICmpInst &I,
...
   }
 
+  // (icmp eq/ne (X, -P2), INT_MIN)
+  //	-> (icmp slt/sge X, INT_MIN + P2)
+  if (ICmpInst::isEquality(Pred) && BO0 &&
+      match(I.getOperand(1), m_SignMask())) {
+    Value *X;
+    if (match(BO0, m_And(m_Value(X), m_CheckedInt([](const APInt &C) {
+                           return C.isZero() || C.isNegatedPowerOf2();
+                         })))) {
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5025]: Use `m_NegatedPower2OrZero` instead?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5025]: Since that matches a splat I cant imagine the zero is necessary.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5025]: m_NegatedPower2OrZero should also match non-splat (the variant without APInt argument).

**Review by nikic — APPROVED** [review]
LG

**Comment by vitalybuka** [comment]
@eugenis 

Looks like this breaks msan
https://lab.llvm.org/buildbot/#/builders/164/builds/3279

And I mean msan itself, not the code.
Similar bot on released compiler, and same code, does not repor https://lab.llvm.org/buildbot/#/builders/169

**Comment by goldsteinn** [comment]
> @eugenis
> 
> Looks like this breaks msan https://lab.llvm.org/buildbot/#/builders/164/builds/3279
> 
> And I mean msan itself, not the code. Similar bot on released compiler, and same code, does not repor https://lab.llvm.org/buildbot/#/builders/169

Hmm? As in MSAN no longer works correctly when compiled with this PR?

**Comment by vitalybuka** [comment]
> > @eugenis
> > Looks like this breaks msan https://lab.llvm.org/buildbot/#/builders/164/builds/3279
> > And I mean msan itself, not the code. Similar bot on released compiler, and same code, does not repor https://lab.llvm.org/buildbot/#/builders/169
> 
> Hmm? As in MSAN no longer works correctly when compiled with this PR?

Yes, I bisected locally..
From what I see reports looks wrong https://lab.llvm.org/buildbot/#/builders/164/builds/3279/steps/14/logs/stdio

**Comment by goldsteinn** [comment]
> > > @eugenis
> > > Looks like this breaks msan https://lab.llvm.org/buildbot/#/builders/164/builds/3279
> > > And I mean msan itself, not the code. Similar bot on released compiler, and same code, does not repor https://lab.llvm.org/buildbot/#/builders/169
> > 
> > 
> > Hmm? As in MSAN no longer works correctly when compiled with this PR?
> 
> Yes, I bisected locally.. Reports looks wrong https://lab.llvm.org/buildbot/#/builders/164/builds/3279/steps/14/logs/stdio

Okay, might just be this PR is buggy and causes miscompiles, although its pretty simple and I don't see it.

**Comment by eugenis** [comment]
This is a problem with MSan handling of signed comparison - it is not entirely correct as a compromise for performance and code size. Do you mind reverting the change while we figure out how to fix MSan?

**Comment by eugenis** [comment]
I can confirm that -msan-handle-icmp-exact=1 avoids this issue, and increases .text size of the clang binary by 2.9%.

**Comment by eugenis** [comment]
https://github.com/llvm/llvm-project/issues/111212 to fix MSan.

**Pull request closed without merging**